File: js/utils/media/loc.ts
export type LocHeader = {
  extensions: LocHeaderExtension[]
}

export type LocHeaderExtension =
  | { type: 'captureTimestamp'; value: { microsSinceUnixEpoch: number } }
  | { type: 'videoConfig'; value: { data: Uint8Array } }
  | { type: 'videoFrameMarking'; value: { data: Uint8Array } }
  | { type: 'audioLevel'; value: { level: number } }
  | { type: 'unknown'; value: { id: number; value: LocHeaderValue } }

export type LocHeaderValue = { evenBytes: Uint8Array } | { oddVarint: number }

export type LocMetadata = {
  captureTimestampMicros?: number
  videoConfig?: Uint8Array
  videoFrameMarking?: Uint8Array
  audioLevel?: number
}

export function buildLocHeader(meta: LocMetadata): LocHeader {
  const extensions: LocHeaderExtension[] = []
  if (typeof meta.captureTimestampMicros === 'number') {
    extensions.push({
      type: 'captureTimestamp',
      value: { microsSinceUnixEpoch: meta.captureTimestampMicros }
    })
  }
  if (meta.videoConfig) {
    extensions.push({
      type: 'videoConfig',
      value: { data: meta.videoConfig }
    })
  }
  if (meta.videoFrameMarking) {
    extensions.push({
      type: 'videoFrameMarking',
      value: { data: meta.videoFrameMarking }
    })
  }
  if (typeof meta.audioLevel === 'number') {
    extensions.push({
      type: 'audioLevel',
      value: { level: meta.audioLevel }
    })
  }
  return { extensions }
}

export function readLocHeader(header?: LocHeader): LocMetadata {
  if (!header) {
    return {}
  }
  const meta: LocMetadata = {}
  for (const ext of header.extensions) {
    switch (ext.type) {
      case 'captureTimestamp':
        meta.captureTimestampMicros = ext.value.microsSinceUnixEpoch
        break
      case 'videoConfig':
        meta.videoConfig = ext.value.data
        break
      case 'videoFrameMarking':
        meta.videoFrameMarking = ext.value.data
        break
      case 'audioLevel':
        meta.audioLevel = ext.value.level
        break
      default:
        break
    }
  }
  return meta
}

export function arrayBufferToUint8Array(buffer?: ArrayBuffer): Uint8Array | undefined {
  if (!buffer) {
    return undefined
  }
  return new Uint8Array(buffer)
}

export function bytesToBase64(bytes: Uint8Array): string {
  let binary = ''
  for (let i = 0; i < bytes.length; i++) {
    binary += String.fromCharCode(bytes[i])
  }
  return btoa(binary)
}

File: moqt-client-rust/src/loc.rs
use anyhow::Result;
use moqt_core::messages::data_streams::extension_header::{
    ExtensionHeader, ExtensionHeaderValue, Value, ValueWithLength,
};
use packages::loc::{
    LocHeader, LocHeaderExtension, LocHeaderValue, LOC_AUDIO_LEVEL_ID, LOC_CAPTURE_TIMESTAMP_ID,
    LOC_VIDEO_CONFIG_ID, LOC_VIDEO_FRAME_MARKING_ID,
};

pub fn loc_header_to_extension_headers(header: &LocHeader) -> Result<Vec<ExtensionHeader>> {
    let mut headers = Vec::with_capacity(header.extensions.len());
    for ext in &header.extensions {
        match ext {
            LocHeaderExtension::CaptureTimestamp(value) => {
                let bytes = value.micros_since_unix_epoch.to_be_bytes().to_vec();
                headers.push(ExtensionHeader::new(
                    LOC_CAPTURE_TIMESTAMP_ID,
                    ExtensionHeaderValue::EvenTypeValue(ValueWithLength::new(bytes)),
                )?);
            }
            LocHeaderExtension::VideoConfig(value) => {
                headers.push(ExtensionHeader::new(
                    LOC_VIDEO_CONFIG_ID,
                    ExtensionHeaderValue::EvenTypeValue(ValueWithLength::new(value.data.clone())),
                )?);
            }
            LocHeaderExtension::VideoFrameMarking(value) => {
                headers.push(ExtensionHeader::new(
                    LOC_VIDEO_FRAME_MARKING_ID,
                    ExtensionHeaderValue::EvenTypeValue(ValueWithLength::new(value.data.clone())),
                )?);
            }
            LocHeaderExtension::AudioLevel(value) => {
                headers.push(ExtensionHeader::new(
                    LOC_AUDIO_LEVEL_ID,
                    ExtensionHeaderValue::EvenTypeValue(ValueWithLength::new(vec![value.level])),
                )?);
            }
            LocHeaderExtension::Unknown(value) => {
                let header = match &value.value {
                    LocHeaderValue::EvenBytes(bytes) => ExtensionHeader::new(
                        value.id,
                        ExtensionHeaderValue::EvenTypeValue(ValueWithLength::new(bytes.clone())),
                    )?,
                    LocHeaderValue::OddVarint(varint) => ExtensionHeader::new(
                        value.id,
                        ExtensionHeaderValue::OddTypeValue(Value::new(*varint)),
                    )?,
                };
                headers.push(header);
            }
        }
    }
    Ok(headers)
}
